src/main/java/ch/epfl/biop/imaris/EasyXT.java

```diff
@@ -779,6 +780,103 @@ public static IDataSetPrx getSurfacesDataset(ISurfacesPrx surface, double downsa
+
+        // Get the whole surface as ImagePlus (0 or 1 ), 8-bit image
+        ImagePlus label_imp = getImagePlus(getSurfacesDataset(surface));
+        if (n_surf > 255 ) IJ.run(label_imp, "16-bit", "");
+        if (n_surf > 65535 ) IJ.run(label_imp, "32-bit", "");
+
+        ArrayList<ImagePlus> imps = new ArrayList<ImagePlus>(last_timepoint);
+
+        // we extract the first timepoint
+        int previous_t=0;
+        ImagePlus t_label_imp = new Duplicator().run(label_imp, 1, 1, 1, label_imp.getNSlices(), previous_t+1, previous_t+1);
+
+        // next will mutiply each sub-surface of the surface, by a int value
+        for (int srf = 0; srf < n_surf; srf++) {
+            // should be final for the processor step below
+            final int val = srf;
+
+            // if the current spot is from a different time-point
+            if ( surface.GetTimeIndex(srf) != previous_t) {
+                // store the current status of t_label_imp into the ArrayList<ImagePlus>
+                // N.B. duplicate is required to store the current time-point
+                imps.add(new ImagePlus("t" + previous_t, t_label_imp.getStack().duplicate()));
+                // ... increment previous_t value
+                previous_t++;
+                // and set the t_label_imp to the next time-point
+                t_label_imp = new Duplicator().run(label_imp, 1, 1, 1, label_imp.getNSlices(), previous_t+1, previous_t+1);
+            }
+
+            // First we tried to use CopySurfaces(id), but it duplicates the Surface
+            // (visible with color based on statistics that make surface display to crash)
+            // int[] id = {i};
+            // ISurfacesPrx current_surface = surface.CopySurfaces(id);
+            // IDataSetPrx dataset = getSurfacesDataset(current_surface);
+
+            // here we make a copy of the whole stack
+            // To go faster we could use a smaller mask by defining the boundingBox of each sub-surface
+            // (but it will also be needed to at the multiply step below)
+            IDataSetPrx current_dataset = surface.GetSingleMask(srf,
+                    (float) cal.xOrigin , (float)cal.yOrigin, (float)cal.zOrigin,
+                    (float) cal.xEnd, (float) cal.yEnd, (float) cal.zEnd,
+                    cal.xSize, cal.ySize, cal.zSize);
+            ImagePlus current_imp = getImagePlus(current_dataset);
+
+            // Multiply by val to get a Label
+            // val could could be replaced by the surface-Imaris-ID )
+            int nProcessor = current_imp.getStack().getSize();
+            IntStream.range(0, nProcessor).parallel().forEach(index -> {
+                current_imp.getStack().getProcessor(index + 1).multiply(val);
+            });
+
+            //now we add the the current_imp to the global t_label_imp
+            // TODO take care of interface between touching objects
+            ImageCalculator ic = new ImageCalculator();
```

Comment:
can be taken outside of loop. No need to declare a new one each time

Reply:
15aac4189009d70be8e1666aea208b7aefab532a

Reply:
Perfect